Production/src/BaseTupleProducer.cc

```diff
@@ -439,7 +455,11 @@ void BaseTupleProducer::FillGenParticleInfo()
     std::vector<const reco::GenParticle*> particles_to_store;
 
     std::map<int, size_t> particle_counts;
-    for(const auto& particle : *genParticles) {
+    for(unsigned n = 0; n < genParticles->size(); ++n) {
+        const auto& particle = genParticles->at(n);
+        if(saveGenParticleInfo){
```

Comment:
don't copy each gen particle in the vector. it is slow. do two loops depending on saveGenParticleInfo and extract the push_back part into a lambda-function.